File: qqzeng-phone-6.0/netcore/PhoneSearch6Db.cs
using System.Buffers.Binary;
using System.Numerics;
using System.Text;

namespace qqzeng_phone_db
{
    public sealed class  PhoneSearch6Db
    {
        private static readonly Lazy<PhoneSearch6Db> _lazy = new(() => new PhoneSearch6Db());
        public static PhoneSearch6Db Instance => _lazy.Value;

        // 常量定义（根据文件格式规范）
        private const int HeaderSize = 32;                  // 8个uint的头部
        private const int PrefixCount = 200;                 // 电话号码前缀总数（0-199）
        private const int BitmapPopCountOffset = 0x4E2;      // 位图统计信息偏移量

        // 字段定义
        private byte[] _data = Array.Empty<byte>();          // 数据库原始数据
        private string[] _regionIsps = Array.Empty<string>(); // 地区-运营商组合缓存
        private readonly (int BitmapOffset, int DataOffset)[] _index
            = new (int, int)[PrefixCount];                   // 前缀索引表


        private PhoneSearch6Db() => LoadDatabase();

        /// <summary>
        /// 加载并解析数据
        /// </summary>
        /// <exception cref="InvalidDataException">文件格式异常</exception>
        public void LoadDatabase()
        {
            var filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "qqzeng-phone-6.0.db");
            if (!File.Exists(filePath)) throw new FileNotFoundException("Database file not found", filePath);
            try
            {
                _data = File.ReadAllBytes(filePath);
                var span = _data.AsSpan();

                // 解析头部（小端序）
                var header = new uint[8];
                for (int i = 0; i < header.Length; i++)
                {
                    header[i] = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(i * 4, 4));
                }

                // 解析地区与运营商表
                int regionsStart = HeaderSize;
                int ispsStart = regionsStart + (int)header[1];
                int indexStart = ispsStart + (int)header[2];

                string[] regions = Encoding.UTF8.GetString(span.Slice(regionsStart, (int)header[1])).Split('&');
                string[] isps = Encoding.UTF8.GetString(span.Slice(ispsStart, (int)header[2])).Split('&');

                // 构建地区-运营商组合
                _regionIsps = new string[header[4]];
                int entryOffset = (int)header[3];
                for (int i = 0; i < _regionIsps.Length; i++)
                {
                    ushort entry = BinaryPrimitives.ReadUInt16LittleEndian(span.Slice(entryOffset + i * 2, 2));
                    _regionIsps[i] = $"{regions[entry >> 5]}|{isps[entry & 0x1F]}";
                }

                // 构建前缀索引表
                int pos = indexStart;
                for (int i = 0; i < PrefixCount; i++)
                {

                    uint prefix = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(pos, 4));
                    if (prefix == i)
                    {
                        _index[i] = (
                            BitmapOffset: BinaryPrimitives.ReadInt32LittleEndian(span.Slice(pos + 4, 4)),
                            DataOffset: BinaryPrimitives.ReadInt32LittleEndian(span.Slice(pos + 8, 4))
                            );
                        pos += 12;
                    }
                    else
                    {
                        _index[i] = (0, 0);
                    }

                }
            }
            catch (Exception ex) when (ex is not FileNotFoundException)
            {
                throw new InvalidDataException("Invalid database format", ex);
            }

        }

        /// <summary>
        /// 查询电话号码归属地信息
        /// </summary>
        /// <param name="phone">7位数字电话号码</param>
        /// <returns>地区|运营商 组合字符串，未找到时返回null</returns>
        /// <exception cref="ArgumentException">无效电话号码格式</exception>
        public string? Query(ReadOnlySpan<char> phone)
        {          

            // 解析前缀和后四位
            int prefix = ParsePhoneSegment(phone[..3]);
            int subNum = ParsePhoneSegment(phone.Slice(3, 4));

          
            // 前缀有效性检查
            if (prefix is < 0 or >= PrefixCount) return null;

            // 获取索引条目
            var (bitmapOffset, dataOffset) = _index[prefix];
            if (bitmapOffset == 0 || dataOffset == 0) return null;

            // 位图检查
            Span<byte> dataSpan = _data.AsSpan();
            int byteIndex = subNum >> 3;        // 等价于 subNum / 8
            int bitIndex = subNum & 0b0111; // 等价于 subNum % 8

            if (bitmapOffset + byteIndex >= dataSpan.Length) return null;

            byte bitmap = dataSpan[bitmapOffset + byteIndex];
            if ((bitmap & (1 << bitIndex)) == 0) return null;

            // 计算有效数据位置
            int popCountOffset = bitmapOffset + BitmapPopCountOffset + (byteIndex << 1);
           
            int preCount = BinaryPrimitives.ReadUInt16LittleEndian(dataSpan.Slice(popCountOffset, 2));
            int localCount = BitOperations.PopCount((uint)(bitmap & ((1 << bitIndex) - 1)));
          

            // 定位最终数据
            int dataPos = dataOffset + ((preCount + localCount) << 1);
            ushort entry = BinaryPrimitives.ReadUInt16LittleEndian(dataSpan.Slice(dataPos, 2));
            return entry < _regionIsps.Length ? _regionIsps[entry] : null;
        }

        /// <summary>
        /// 将电话号码段转换为数字
        /// </summary>
        private static int ParsePhoneSegment(ReadOnlySpan<char> segment)
        {
            int result = 0;
            foreach (char c in segment)
            {
                result = result * 10 + (c - '0');
            }
            return result;
        }
    }
    /*
     *   var local = PhoneSearch6Db.Instance.Query(phone);
     *   return 省份|城市|邮编|区号|行政区划代码|运营商
     *   
    查询 qqzeng-phone-6.0.db 7598万 ->1.484秒 每秒5119.946091644205万次
    查询 qqzeng-phone-6.0.db 6380万 ->1.075秒 每秒5934.883720930233万次
    查询 qqzeng-phone-6.0.db 6438万 ->1.102秒 每秒5842.105263157894万次
    查询 qqzeng-phone-6.0.db 10904万 ->1.889秒 每秒5772.366331392271万次
    查询 qqzeng-phone-6.0.db 9454万 ->1.584秒 每秒5968.434343434343万次
    查询 qqzeng-phone-6.0.db 9048万 ->1.466秒 每秒6171.896316507504万次
    查询 qqzeng-phone-6.0.db 7598万 ->1.219秒 每秒6232.9778506972925万次
    */


}

File: phone-location-sdk/v6.0/nodejs/PhoneSearch6Db.js
const fs = require('fs');
const path = require('path');

const HeaderSize = 32;
const PrefixCount = 200;
const BitmapPopCountOffset = 0x4E2;

class PhoneSearch6Db {
  constructor() {
    this.data = Buffer.alloc(0); // 用于存储数据库文件的二进制数据
    this.regionIsps = []; // 地区-运营商组合
    this.index = new Array(PrefixCount).fill({ BitmapOffset: 0, DataOffset: 0 });
    this.loadDatabase();
  }

  // 加载并解析数据库文件
  loadDatabase() {
    const filePath = path.join(__dirname, 'qqzeng-phone-6.0.db');
    if (!fs.existsSync(filePath)) throw new Error(`Database file not found: ${filePath}`);

    try {
      // 读取数据库文件为 Buffer
      this.data = fs.readFileSync(filePath);
      const span = this.data;

      // 解析头部（小端序）
      const header = [];
      for (let i = 0; i < 8; i++) {
        header[i] = span.readUInt32LE(i * 4);
      }

      // 解析地区与运营商表
      const regionsStart = HeaderSize;
      const ispsStart = regionsStart + header[1];
      const indexStart = ispsStart + header[2];

      const regions = span.toString('utf8', regionsStart, regionsStart + header[1]).split('&');
      const isps = span.toString('utf8', ispsStart, ispsStart + header[2]).split('&');

      // 构建地区-运营商组合
      this.regionIsps = new Array(header[4]);
      const entryOffset = header[3];
      for (let i = 0; i < this.regionIsps.length; i++) {
        const entry = span.readUInt16LE(entryOffset + i * 2);
        this.regionIsps[i] = `${regions[entry >> 5]}|${isps[entry & 0x1F]}`;
      }

      // 构建前缀索引表
      let pos = indexStart;
      for (let i = 0; i < PrefixCount; i++) {
        const prefix = span.readUInt32LE(pos);
        if (prefix === i) {
          this.index[i] = {
            BitmapOffset: span.readUInt32LE(pos + 4),
            DataOffset: span.readUInt32LE(pos + 8),
          };
          pos += 12;
        }
      }
    } catch (err) {
      throw new Error(`Invalid database format: ${err.message}`);
    }
  }

  // 查询电话号码归属地信息
  query(phone) {
    if (phone.length !== 7) {
      throw new Error('Invalid phone number format');
    }

    // 解析前缀和后四位
    const prefix = this.parsePhoneSegment(phone.slice(0, 3));
    const subNum = this.parsePhoneSegment(phone.slice(3, 7));

    // 前缀有效性检查
    if (prefix < 0 || prefix >= PrefixCount) return null;

    // 获取索引条目
    const { BitmapOffset, DataOffset } = this.index[prefix];
    if (BitmapOffset === 0 || DataOffset === 0) return null;

    // 位图检查
    const byteIndex = subNum >> 3;
    const bitIndex = subNum & 0b0111;

    if (BitmapOffset + byteIndex >= this.data.length) return null;

    const bitmap = this.data.readUInt8(BitmapOffset + byteIndex);
    if ((bitmap & (1 << bitIndex)) === 0) return null;

    // 计算有效数据位置
    const popCountOffset = BitmapOffset + BitmapPopCountOffset + (byteIndex << 1);
    const preCount = this.data.readUInt16LE(popCountOffset);
    const localCount = this.countSetBits(bitmap & ((1 << bitIndex) - 1));

    // 定位最终数据
    const dataPos = DataOffset + ((preCount + localCount) << 1);
    const entry = this.data.readUInt16LE(dataPos);

    return entry < this.regionIsps.length ? this.regionIsps[entry] : null;
  }

  // 将电话号码段转换为数字
  parsePhoneSegment(segment) {
    let result = 0;
    for (const char of segment) {
      result = result * 10 + (char.charCodeAt(0) - '0'.charCodeAt(0));
    }
    return result;
  }

  // 计算有效位的数量（popcount）
  countSetBits(bitmap) {
    let count = 0;
    while (bitmap > 0) {
      count++;
      bitmap &= bitmap - 1;
    }
    return count;
  }
}

module.exports = PhoneSearch6Db;
